dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -0,0 +1,265 @@
+        resource_group_task_queues.insert({name, task_queue});
+    }
+    else
+    {
+        iter->second->submit(std::move(task));
+    }
+    cv.notify_one();
+}
+
+template <typename NestedQueueType>
+bool ResourceControlQueue<NestedQueueType>::take(TaskPtr & task)
+{
+    assert(task == nullptr);
+    std::unique_lock lock(mu);
+    while (true)
+    {
+        if (popTask(cancel_task_queue, task))
+            return true;
+
+        while (!resource_group_infos.empty())
+        {
+            if unlikely (is_finished)
+            {
+                // resource_group_infos and resource_group_task_queues will be cleaned in destructor,
+                // and all Tasks in nested task queue will be drained in destructor of TaskQueue.
+                return false;
+            }
+
+            ResourceGroupInfo group_info = resource_group_infos.top();
+            const std::string & name = std::get<InfoIndexResourceGroupName>(group_info);
+            const auto & priority = LocalAdmissionController::global_instance->getPriority(name);
+            std::shared_ptr<NestedQueueType> & task_queue = std::get<InfoIndexPipelineTaskQueue>(group_info);
+
+            LOG_TRACE(
+                logger,
+                "trying to schedule task of resource group {}, priority: {}, is_finished: {}, task_queue.empty(): {}",
+                name,
+                priority,
+                is_finished,
+                task_queue->empty());
+
+            // When highest priority of resource group is less than zero, means RU of all resource groups are exhausted.
+            // Should not take any task from nested task queue for this situation.
+            if (LocalAdmissionController::isRUExhausted(priority))
+                break;
+
+            if (task_queue->empty() || !task_queue->take(task))
+            {
+                LOG_TRACE(
+                    logger,
+                    "take task from nested task_queue of resource group {} failed. task_queue.empty(): {}",
+                    name,
+                    task_queue->empty());
+                // Got here only when task_queue is empty or finished, we try next resource group.
+                // If new task of this resource gorup is submited, the resource_group info will be added again.
+                resource_group_infos.pop();
+                size_t erase_num = resource_group_task_queues.erase(name);
+                RUNTIME_CHECK_MSG(
+                    erase_num == 1,
+                    "cannot erase corresponding TaskQueue for task of resource group {}",
+                    name);
+            }
+            else
+            {
+                LOG_TRACE(
+                    logger,
+                    "schedule task of resource group {} succeed, cur cpu time of MPPTask: {}",
+                    name,
+                    task->getQueryExecContext().getQueryProfileInfo().getCPUExecuteTimeNs());
```

Comment:
Getting `task->getQueryExecContext().getQueryProfileInfo().getCPUExecuteTimeNs()` is meaningless, because this value will not be updated at runtime, only when the task ends

Reply:
Or `task->profile_info.getCPUExecuteTimeNs()`?

Reply:
ok